File: frontend/src/components/bank/bankUndo.js
/** ↩ Undo the last bulk decision — pure helpers (no JSX, so `node --test`
 * can run them).
 *
 * Marking hundreds of images in one gesture is the bank's biggest lever. The
 * backend now snapshots the (status, reason) of every row a bulk action flips
 * and offers ONE step back, per bank, in the payload the workspace already
 * polls (`payload.undo` = {label, count, at}) — which is why the bar survives a
 * reload: the decision it takes back lives in the database, not in this tab.
 *
 * These helpers own the WORDING, and the wording is where the honesty lives.
 * Two rules run through the whole file:
 *   1. never promise what the server did not do — a restore that put 340 of 400
 *      rows back says so, and names what it left alone;
 *   2. never offer an undo for an action that cannot be undone cleanly. The
 *      server simply publishes no offer for Delete rejected or ⬆ Promote,
 *      so there is nothing here to accidentally render.
 */

/** The offer to show, or null. Defensive about the payload shape: a count of 0
 * is "nothing was actually flipped", which must not draw a bar. */
export function undoOffer(payload) {
  const undo = payload?.undo
  if (!undo) return null
  const count = Number(undo.count) || 0
  if (count <= 0) return null
  const label = String(undo.label || 'Last bulk action')
  return { label, count, at: Number(undo.at) || 0 }
}

/** The bar's sentence. Kept to one line at 400 px: the action, how many images
 * it moved, and the boundary of the promise (this session only). */
export function undoBannerText(offer) {
  if (!offer) return ''
  return `${offer.label} — ${offer.count} image${offer.count === 1 ? '' : 's'}`
}

/** The tooltip/second line: what pressing it will do, and what it will not.
 * Users deserve the restart caveat BEFORE they rely on the net, not after. */
export const UNDO_HINT = 'Puts those images back exactly as they were. '
  + 'One step only, and only until the app restarts.'

/** The honest ledger after a restore. Never a bare "done": every image that did
 * not make it back is counted, and the ones a newer decision now owns are named
 * so the user can go and look at them. */
export function undoResultMessage(result) {
  if (!result) return { type: 'error', text: 'Undo failed.' }
  const restored = Number(result.restored) || 0
  const missing = Number(result.missing) || 0
  const conflicts = Number(result.conflicts) || 0
  const total = Number(result.total) || restored + missing + conflicts
  const names = Array.isArray(result.conflict_names) ? result.conflict_names : []

  if (restored === total && !missing && !conflicts) {
    return {
      type: 'success',
      text: `↩ Restored ${restored} image${restored === 1 ? '' : 's'} to what they were.`,
    }
  }
  const parts = [`↩ Restored ${restored} of ${total} images.`]
  if (missing) {
    parts.push(`${missing} ${missing === 1 ? 'is' : 'are'} no longer in the bank.`)
  }
  if (conflicts) {
    const shown = names.slice(0, 3).join(', ')
    parts.push(`${conflicts} changed since and ${conflicts === 1 ? 'was' : 'were'} left alone`
      + (shown ? ` (${shown}${conflicts > names.slice(0, 3).length ? ', …' : ''}).` : '.'))
  }
  return { type: restored ? 'info' : 'error', text: parts.join(' ') }
}

File: backend/app/services/bank_undo.py
"""↩ Undo for the bank's bulk decisions — one step back, per bank.

WHY THIS EXISTS
---------------
Marking hundreds of images in one gesture (✓/✕ over a whole filter, auto-reject
at a threshold, collapsing every duplicate group) is the bank's biggest lever and
was its only irreversible-feeling one. A mis-set threshold used to cost a manual
re-triage of everything it caught.

WHAT IT REMEMBERS — and what it deliberately does not
-----------------------------------------------------
Only the STATUS DIMENSION: ``(status, reject_reason)`` per row, the two columns
those actions write. That is exactly the set of bulk actions that are reversible
*cleanly*, and the honesty rule is that we offer undo for nothing else:

* 🗑 Delete rejected sends source files to the OS trash (which we cannot pull
  back programmatically) and drops the rows with all their analysis — an
  "Undo" there would restore, at best, half of half. It CLEARS this registry
  instead, because the rows a pending snapshot points at are the ones it just
  deleted.
* ⬆ Promote copies through the dataset import path; un-promoting would mean
  deleting images in someone else's dataset.
* 🔄 Rotate lives on another column and already has a natural inverse (turn the
  other way), which the toast says.

DEPTH IS ONE, ON PURPOSE
------------------------
Not a shortcut — the only depth at which "every row goes back to exactly what it
was" stays checkable. A deeper stack would have to reason about whether older
states still hold after a re-scan dropped rows or a pipeline re-flagged them, and
would invite unwinding *through* a pipeline whose other effects do not unwind.
Zero → one is the whole value.

IN MEMORY, LIKE THE JOB REGISTRY
--------------------------------
Same declared lifetime as ``bank_jobs`` and ``dataset_activity``: the snapshot
lives in the SERVER, so it survives a page reload, a tab switch and a second tab
— which a React-state undo does not — and dies with the process. A restart is a
defensible session boundary for "undo my last click", and it costs no schema
migration on the databases users already have. The UI states the boundary rather
than implying permanence.

A snapshot entry is ``{image_id: {'before': (status, reason),
'after': (status, reason)}}``. Keeping ``after`` is what lets the restore tell
"nobody touched this since" from "someone did" — the latter is skipped and
NAMED, never clobbered.
"""
import threading
import time

_lock = threading.Lock()
_snapshots: dict = {}        # bank_id -> {'label', 'rows', 'at'}
_TTL = 60 * 60               # an offer older than this is stale — the user moved on


class Snapshot:
    """Collects the prior state of the rows an action is about to change.

    Mutators call :meth:`note` right before writing; rows whose value would not
    actually change are ignored, so the offer's count is the number of decisions
    genuinely flipped rather than the size of the selection.

    One snapshot can span several mutators (the pipeline's auto-reject step is
    ``apply_flags`` followed by a duplicate resolution): the earliest ``before``
    wins and the latest ``after`` is kept, so the pair still describes the whole
    step end to end.
    """

    def __init__(self, label):
        self.label = label
        self.rows: dict = {}

    def note(self, row, status, reason):
        before = (row.status, row.reject_reason)
        after = (status, reason)
        entry = self.rows.get(row.id)
        if entry is None:
            if before == after:
                return
            self.rows[row.id] = {'before': before, 'after': after}
            return
        entry['after'] = after
        if entry['before'] == after:
            # walked back to where it started within the same action
            self.rows.pop(row.id, None)

    def commit(self, bank_id):
        """Publish as THE undo offer for this bank, replacing any previous one.
        A snapshot that changed nothing publishes nothing."""
        if self.rows:
            record(bank_id, self.label, self.rows)
        return len(self.rows)


def record(bank_id, label, rows):
    """Replace the bank's undo offer. ``rows`` is {image_id: {'before','after'}}."""
    with _lock:
        _snapshots[int(bank_id)] = {'label': str(label), 'rows': dict(rows),
                                    'at': time.time()}


def peek(bank_id):
    """The offer as the UI shows it: {label, count, at} — or None. Purges a
    stale entry on the way through, like the job registry does."""
    with _lock:
        snap = _snapshots.get(int(bank_id))
        if not snap:
            return None
        if time.time() - snap['at'] > _TTL:
            _snapshots.pop(int(bank_id), None)
            return None
        return {'label': snap['label'], 'count': len(snap['rows']),
                'at': snap['at']}


def take(bank_id):
    """Pop the offer for a restore. One shot: an undo that half-succeeded is not
    re-offered, because a second press would find the same conflicts."""
    with _lock:
        snap = _snapshots.pop(int(bank_id), None)
    if snap and time.time() - snap['at'] > _TTL:
        return None
    return snap


def clear(bank_id):
    """Drop the offer — used by the actions that make it un-restorable
    (🗑 Delete rejected drops the very rows it points at) and by bank deletion."""
    with _lock:
        _snapshots.pop(int(bank_id), None)


def reset():
    """Test helper: forget every snapshot."""
    with _lock:
        _snapshots.clear()
